File: android/app/src/main/java/com/flashairsync/core/SyncEngine.kt
package com.flashairsync.core

import android.content.Context
import android.net.Network
import android.os.Build
import android.util.Log
import androidx.annotation.RequiresApi
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.withContext
import java.io.File

/**
 * Main sync orchestrator
 * Coordinates WiFi connection, file discovery, download, and MediaStore saving
 */
@RequiresApi(Build.VERSION_CODES.Q)
class SyncEngine(
    private val context: Context,
    private val settings: SyncSettings
) {

    companion object {
        private const val TAG = "SyncEngine"
        private const val MAX_RETRIES = 3
        private const val RETRY_DELAY_MS = 2000L
    }

    private val wifiConnector = WiFiConnector(context)
    private val syncIndex = SyncIndex(context, settings.ssid)
    private val mediaStoreWriter = MediaStoreWriter(context)

    private val _syncState = MutableStateFlow<SyncState>(SyncState.Idle)
    val syncState: StateFlow<SyncState> = _syncState.asStateFlow()

    private val _importItems = MutableStateFlow<List<ImportItem>>(emptyList())
    val importItems: StateFlow<List<ImportItem>> = _importItems.asStateFlow()

    private var isCancelled = false

    // MARK: - Public API

    /**
     * Start sync operation
     * @param rootPath Starting directory (e.g., "/DCIM")
     * @return ImportResult with statistics
     */
    suspend fun startSync(rootPath: String = "/DCIM"): ImportResult = withContext(Dispatchers.IO) {
        isCancelled = false
        val startTime = System.currentTimeMillis()

        try {
            // Step 1: Connect to WiFi
            _syncState.value = SyncState.ConnectingWiFi
            Log.d(TAG, "Connecting to WiFi: ${settings.ssid}")

            val network = wifiConnector.connect(settings.ssid, settings.passphrase)
            _syncState.value = SyncState.WiFiConnected(network)

            // Step 2: Discover files
            _syncState.value = SyncState.Discovering
            Log.d(TAG, "Discovering files in $rootPath")

            val client = FlashAirClient(settings.host)
            val allEntries = client.walkDirectory(rootPath) { entry ->
                settings.shouldImport(entry)
            }

            Log.d(TAG, "Found ${allEntries.size} eligible files")

            // Step 3: Filter out already-synced files
            val newEntries = allEntries.filter { !syncIndex.hasSeen(it) }
            Log.d(TAG, "New files to import: ${newEntries.size}, Already synced: ${allEntries.size - newEntries.size}")

            if (newEntries.isEmpty()) {
                _syncState.value = SyncState.Completed(
                    ImportResult(
                        totalFiles = 0,
                        importedCount = 0,
                        skippedCount = allEntries.size,
                        failedCount = 0,
                        totalBytes = 0,
                        duration = System.currentTimeMillis() - startTime,
                        errors = emptyList()
                    )
                )

                // Disconnect WiFi
                wifiConnector.disconnect()

                return@withContext _syncState.value.let { it as SyncState.Completed }.result
            }

            // Step 4: Create import items
            val items = newEntries.map { ImportItem(it) }
            _importItems.value = items
            _syncState.value = SyncState.Downloading(0, items.size)

            // Step 5: Download and save files
            var importedCount = 0
            var failedCount = 0
            var totalBytes = 0L
            val errors = mutableListOf<String>()

            for ((index, item) in items.withIndex()) {
                if (isCancelled) {
                    Log.d(TAG, "Sync cancelled by user")
                    throw FlashAirError.Cancelled
                }

                try {
                    // Update item state
                    updateItemState(item.id, ImportState.Downloading(0f))

                    // Download file with retries
                    val tempFile = downloadFileWithRetry(client, item.entry) { progress ->
                        updateItemState(item.id, ImportState.Downloading(progress))
                    }

                    // Save to MediaStore
                    updateItemState(item.id, ImportState.Saving)
                    val uri = mediaStoreWriter.saveEntry(item.entry, tempFile)
                    Log.d(TAG, "Saved ${item.entry.name} to MediaStore: $uri")

                    // Update sync index
                    syncIndex.markSeen(item.entry)

                    // Mark as completed
                    updateItemState(item.id, ImportState.Completed)
                    importedCount++
                    totalBytes += item.entry.size

                    // Clean up temp file
                    tempFile.delete()

                } catch (e: Exception) {
                    Log.e(TAG, "Failed to import ${item.entry.name}: ${e.message}", e)
                    updateItemState(item.id, ImportState.Failed(e.message ?: "Unknown error"))
                    errors.add("${item.entry.name}: ${e.message}")
                    failedCount++
                }

                // Update progress
                _syncState.value = SyncState.Downloading(index + 1, items.size)
            }

            // Step 6: Persist sync index
            syncIndex.persist()

            // Step 7: Build result
            val result = ImportResult(
                totalFiles = items.size,
                importedCount = importedCount,
                skippedCount = allEntries.size - newEntries.size,
                failedCount = failedCount,
                totalBytes = totalBytes,
                duration = System.currentTimeMillis() - startTime,
                errors = errors
            )

            _syncState.value = SyncState.Completed(result)
            Log.d(TAG, "Sync completed: ${result.summary}")

            result

        } catch (e: Exception) {
            Log.e(TAG, "Sync failed: ${e.message}", e)
            _syncState.value = SyncState.Failed(e.message ?: "Unknown error")
            throw e

        } finally {
            // Always disconnect WiFi
            wifiConnector.disconnect()
        }
    }

    /**
     * Cancel ongoing sync operation
     */
    fun cancel() {
        isCancelled = true
        wifiConnector.disconnect()
        _syncState.value = SyncState.Idle
    }

    /**
     * Reset sync index (clear all seen files)
     */
    suspend fun resetIndex() {
        syncIndex.reset()
    }

    /**
     * Get sync index statistics
     */
    fun getIndexStats(): IndexStats {
        return syncIndex.getStats()
    }

    // MARK: - Private Helpers

    /**
     * Download file with retry logic
     */
    private suspend fun downloadFileWithRetry(
        client: FlashAirClient,
        entry: DirectoryEntry,
        onProgress: (Float) -> Unit
    ): File {
        var lastException: Exception? = null

        repeat(MAX_RETRIES) { attempt ->
            try {
                // Create temp file
                val tempFile = File.createTempFile(
                    "flashair_",
                    "_${entry.name}",
                    context.cacheDir
                )

                // Download
                client.downloadFile(entry.absolutePath, tempFile, onProgress)

                // Verify size
                if (tempFile.length() != entry.size) {
                    tempFile.delete()
                    throw Exception("Size mismatch: expected ${entry.size}, got ${tempFile.length()}")
                }

                return tempFile

            } catch (e: Exception) {
                lastException = e
                Log.w(TAG, "Download attempt ${attempt + 1} failed: ${e.message}")

                if (attempt < MAX_RETRIES - 1) {
                    delay(RETRY_DELAY_MS * (attempt + 1))
                }
            }
        }

        throw lastException ?: Exception("Download failed after $MAX_RETRIES attempts")
    }

    /**
     * Update import item state
     */
    private fun updateItemState(itemId: String, newState: ImportState) {
        val updatedItems = _importItems.value.map { item ->
            if (item.id == itemId) {
                item.copy(state = newState)
            } else {
                item
            }
        }
        _importItems.value = updatedItems
    }
}

// MARK: - Sync State

/**
 * Overall sync operation state
 */
sealed class SyncState {
    object Idle : SyncState()
    object ConnectingWiFi : SyncState()
    data class WiFiConnected(val network: Network) : SyncState()
    object Discovering : SyncState()
    data class Downloading(val current: Int, val total: Int) : SyncState()
    data class Completed(val result: ImportResult) : SyncState()
    data class Failed(val error: String) : SyncState()

    val isActive: Boolean
        get() = this !is Idle && this !is Completed && this !is Failed

    val progressPercent: Float
        get() = when (this) {
            is Downloading -> if (total > 0) current.toFloat() / total else 0f
            is Completed -> 1f
            else -> 0f
        }

    val statusMessage: String
        get() = when (this) {
            is Idle -> "Ready"
            is ConnectingWiFi -> "Connecting to WiFi..."
            is WiFiConnected -> "Connected"
            is Discovering -> "Discovering files..."
            is Downloading -> "Downloading $current of $total files"
            is Completed -> "Sync complete: ${result.importedCount} files imported"
            is Failed -> "Failed: $error"
        }
}

File: ios/FlashAirSync/FlashAirSync/Core/SyncEngine.swift
import Foundation

/// Orchestrates the sync process: scan, dedupe, download
actor SyncEngine {
    private let client: FlashAirClient
    private let index: SyncIndex
    private let settings: SyncSettings
    private let photoSaver = PhotoSaver()

    private(set) var isCancelled = false

    init(settings: SyncSettings) {
        self.settings = settings
        self.client = FlashAirClient(host: settings.host)
        self.index = SyncIndex(ssid: settings.ssid)
    }

    // MARK: - Sync Operations

    /// Perform a full sync operation
    /// - Parameters:
    ///   - progressCallback: Called for each state change
    /// - Returns: Array of import items with final states
    func sync(
        progressCallback: @MainActor @escaping ([ImportItem]) -> Void,
        phaseCallback: @MainActor @escaping (String?) -> Void
    ) async throws -> ImportResult {
        isCancelled = false
        let startTime = Date()

        // Step 1: Scan DCIM directory recursively
        await phaseCallback("Scanning files on the card…")
        print("📡 Scanning /DCIM…")
        await Logger.shared.logInfo("Scanning /DCIM")
        let allEntries = try await client.walkDirectory("/DCIM") { [settings] entry in
            settings.shouldImport(entry)
        }
        print("📊 Found \(allEntries.count) files on card after extension/size filter")
        await Logger.shared.logInfo("Found \(allEntries.count) files after extension/size filter")

        // Step 2: Filter out already-synced files
        await phaseCallback("Checking sync history (\(allEntries.count) on card)…")
        var newEntries: [DirectoryEntry] = []
        for entry in allEntries {
            if await !index.hasSeen(entry) {
                newEntries.append(entry)
            }
        }
        print("🔁 \(allEntries.count - newEntries.count) already synced, \(newEntries.count) new")
        await Logger.shared.logInfo("\(allEntries.count - newEntries.count) already synced, \(newEntries.count) new")

        // Optional per-sync cap (nil = unlimited). Useful for trying the app
        // on a card with thousands of photos without committing to syncing all.
        if let limit = settings.maxFilesPerSync, limit > 0, newEntries.count > limit {
            print("📋 Capping this sync to \(limit) of \(newEntries.count) new files (maxFilesPerSync setting)")
            newEntries = Array(newEntries.prefix(limit))
        }

        print("✨ \(newEntries.count) new files to import")

        // Step 3: Create import items
        var items = newEntries.map { ImportItem(entry: $0, state: .pending) }
        await progressCallback(items)

        // Hand the header off to the per-file count ("Importing X of N…")
        // that statusText computes from items during the download loop.
        await phaseCallback(newEntries.isEmpty ? "Nothing new to import" : nil)

        // Step 4: Download and import each file
        var totalBytes = 0
        var importedCount = 0
        var failedCount = 0
        var errors: [String] = []

        for i in 0..<items.count {
            if isCancelled {
                items[i].state = .skipped(reason: "Cancelled")
                continue
            }

            let entry = items[i].entry

            do {
                // Update state: downloading
                items[i].state = .downloading(progress: 0.0)
                await progressCallback(items)

                // Download file (TODO: add progress tracking)
                let localURL = try await client.downloadFile(entry.absolutePath)

                // Update state: saving
                items[i].state = .saving
                await progressCallback(items)

                // Save to Photos. saveToPhotos throws on failure, so the
                // markSeen below is only reached on a successful save — a
                // failed file stays un-indexed and will retry on next sync.
                try await photoSaver.saveToPhotos(
                    fileURL: localURL,
                    originalFilename: entry.name,
                    creationDate: entry.modifiedAt
                )

                // Mark as imported (only after the file is safely in Photos)
                await index.markSeen(entry)
                totalBytes += entry.size
                importedCount += 1

                // Clean up temp file
                try? FileManager.default.removeItem(at: localURL)

                // Update state: completed
                items[i].state = .completed
                await progressCallback(items)

                print("✅ Imported: \(entry.name) (\(ByteCountFormatter.string(fromByteCount: Int64(entry.size), countStyle: .file)))")

            } catch {
                failedCount += 1
                let errorMsg = error.localizedDescription
                errors.append("\(entry.name): \(errorMsg)")

                items[i].state = .failed(error: errorMsg)
                await progressCallback(items)

                print("❌ Failed: \(entry.name) - \(errorMsg)")
            }
        }

        // Step 5: Persist index
        try await index.persist()

        let duration = Date().timeIntervalSince(startTime)

        return ImportResult(
            totalFiles: items.count,
            importedCount: importedCount,
            skippedCount: items.count - importedCount - failedCount,
            failedCount: failedCount,
            totalBytes: totalBytes,
            duration: duration,
            errors: errors
        )
    }

    /// Cancel the current sync operation
    func cancel() {
        isCancelled = true
    }

    /// Get sync index statistics
    func getIndexStats() async -> IndexStats {
        await index.getStats()
    }

    /// Reset sync index (re-import all files on next sync)
    func resetIndex() async {
        await index.reset()
    }
}
